File: back-end/file_types.py
"""The single list of file extensions this app accepts, and what reads each one.

Five independent whitelists used to disagree, so what a file did depended on which
door it came through:
  - `.doc` was accepted everywhere and readable nowhere.
  - `.xlsx/.xls/.csv` were accepted by the upload route but rejected by Generate.
  - `.tiff` was accepted only by Generate.
  - `.pptx/.txt` were accepted only by Smart Import, whose extractor had no branch
    for either, so every slide deck imported and then failed processing.

Top-level module rather than `app/utils/` on purpose: `config.py` needs it, and
importing anything under `app.` from `config` would run `app/__init__.py`, which
imports `config` right back.
"""

# Rasterised or vector documents that go through a text extractor.
DOCUMENT_EXTENSIONS = frozenset({"pdf", "docx", "doc", "pptx", "txt"})

# Images — read by provider vision OCR.
IMAGE_EXTENSIONS = frozenset({"png", "jpg", "jpeg", "webp", "bmp", "tiff"})

# Tabular sources, including exported question banks.
SPREADSHEET_EXTENSIONS = frozenset({"xlsx", "xls", "csv"})

SUPPORTED_EXTENSIONS = DOCUMENT_EXTENSIONS | IMAGE_EXTENSIONS | SPREADSHEET_EXTENSIONS

# Smart Import walks a directory the user picked, so it deliberately leaves images
# out: pointing it at a folder that happens to contain photos would spend vision
# OCR quota on holiday snaps before anything could be classified.
SMART_IMPORT_EXTENSIONS = DOCUMENT_EXTENSIONS | SPREADSHEET_EXTENSIONS

# Comma-separated `accept` attribute for the browser file picker.
ACCEPT_ATTRIBUTE = ",".join(sorted(f".{e}" for e in SUPPORTED_EXTENSIONS))

File: front-end/src/lib/file-types.ts
/**
 * File types the backend accepts, mirrored for the browser file picker.
 *
 * Kept in one place because three components declared the same `accept` string
 * independently and had already drifted from the backend: the picker offered
 * `.doc` and `.xlsx` but not `.pptx`, `.txt` or `.tiff`, so a user could not
 * choose a slide deck the server was perfectly able to read.
 *
 * Source of truth is `back-end/file_types.py` — keep the two in step.
 */

export const DOCUMENT_EXTENSIONS = [
  "pdf",
  "docx",
  "doc",
  "pptx",
  "txt",
] as const;

export const IMAGE_EXTENSIONS = [
  "png",
  "jpg",
  "jpeg",
  "webp",
  "bmp",
  "tiff",
] as const;

export const SPREADSHEET_EXTENSIONS = ["xlsx", "xls", "csv"] as const;

export const SUPPORTED_EXTENSIONS: readonly string[] = [
  ...DOCUMENT_EXTENSIONS,
  ...IMAGE_EXTENSIONS,
  ...SPREADSHEET_EXTENSIONS,
];

/** Value for an `<input type="file">` accept attribute. */
export const FILE_ACCEPT_ATTRIBUTE = SUPPORTED_EXTENSIONS.map(
  (ext) => `.${ext}`,
).join(",");

/** Mirrors Flask's MAX_CONTENT_LENGTH (config.py). */
export const MAX_UPLOAD_BYTES = 50 * 1024 * 1024;
export const MAX_UPLOAD_MB = MAX_UPLOAD_BYTES / (1024 * 1024);

export function isSupportedFile(name: string): boolean {
  const ext = name.includes(".")
    ? name.slice(name.lastIndexOf(".") + 1).toLowerCase()
    : "";
  return SUPPORTED_EXTENSIONS.includes(ext);
}
